### ISSUE TITLE
Self-supervision/distillation in AllenNLP

### ISSUE THREAD

**Issue by ethch18** [allenai/allennlp#4861]
Hi!  I'm trying to train a model in a self-supervised/distilling manner, in the following setting:
* Model A is a normal AllenNLP model (e.g., biaffine dependency parser) specified with a standard config and trained on standard data
* Model B is an AllenNLP model similar to Model A (e.g., also a biaffine dependency parser, but with some slight changes) trained on the label distributions of Model A
* Model A and Model B should have the same tokenization schemes, dataset readers, and indexers (so that B's inputs can be passed into A)

I'm just starting on this and am mainly trying to get a sense of what would need to be added/modified, so I don't have a MWE yet.  But here's my current thought process of what would need to be done manually, based on my understanding of the library:
* Tokenization/dataset intake can be specified in the dataset reader config
* The vocabularies would need to be somehow tied if I were using some kind of ID-based indexer, but if I use a pretrained *BERT model, this issue is resolved by pinning to the pretrained model's vocabulary
* The saved Model A can be loaded into Model B with `from_archive`

Does this seem complete, or are there additional steps that I'd need to take to do this correctly?  Or is there a better way to do this off-the-shelf in AllenNLP?  Thanks!

**Comment by epwalsh** [comment]
I think you've got it covered.

And for the case when you're NOT using a pretrained transformer model and therefore need another way to ensure the  vocabularies match, you can set the vocab part of Model B's config to: `"vocabulary": {"type": "from_files", "directory": "model_A.tar.gz"}`.

**Comment by ethch18** [comment]
Great, thanks!  Follow up question, is there a way to ingest a dataset with mixed labeling, i.e., have some of the data be unlabeled and some of it labeled?  Ideally, I'd be able to check if the label field was None or not, and use that to determine whether to train on label distributions vs. hard labels.

It seems that this could be done through a custom dataset reader, but I get the sense that this would run into problems when processing batches, since part of a batch might have labels and the other part not.  Do any workarounds come to mind?

**Comment by epwalsh** [comment]
Hmm yea batching would be an issue. We're actually working on making it much easier to do multi-task learning, but that API isn't stable yet (it's currently on the [`vision`](https://github.com/allenai/allennlp/tree/vision) branch). You could try it out though, or at least take a look at it to get some ideas. Basically, the multi-task stuff involves a custom [`DataLoader`](https://github.com/allenai/allennlp/tree/vision/allennlp/data/data_loaders) which knows to only batch instances together that below to the same task, and a [modeling component](https://github.com/allenai/allennlp/blob/vision/allennlp/models/multitask.py) which wraps different "heads" for each task.

Fair warning though: the stuff on the `vision` branch involves some breaking changes, mostly around data loading, and will eventually be shipped in our 2.0 release.

**Comment by ethch18** [comment]
Thanks, Evan!  I'll take a look at these.
